File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def getHeight(self, node):
        if node is None:
            return 0
        leftH = self.getHeight(node.left)
        rightH = self.getHeight(node.right)
        if leftH == -1 or rightH == -1 or abs(leftH - rightH) > 1:
            return -1
        return max(leftH, rightH) + 1
    
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.getHeight(root) != -1
        

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int getHeight(TreeNode* node){
        if (node == nullptr)
            return 0;
        // -1 means unbalanced
        int leftHeight = getHeight(node->left);
        int rightHeight = getHeight(node->right);
        if (leftHeight == -1 || rightHeight == -1)
            return -1;
        return abs(leftHeight - rightHeight) > 1? -1 : max(leftHeight, rightHeight)+1;
    }
    
    bool isBalanced(TreeNode* root) {
        return getHeight(root) == -1 ? false : true;
    }
};